File: examples/nodejs/real_snapshot.ts
import { TsgoApiClient } from "@corsa-bind/node";

import { assertExists, isMain, realBinary, realDataset, workspaceRoot } from "../shared.ts";

export function runRealSnapshotExample() {
  assertExists(realBinary, "real tsgo binary", "run `vp run -w build_tsgo` first");
  assertExists(realDataset, "pinned tsgo dataset", "run `vp run -w sync_origin` first");

  const client = TsgoApiClient.spawn({
    executable: realBinary,
    cwd: workspaceRoot,
    mode: "msgpack",
  });

  let snapshotHandle: string | undefined;

  try {
    client.initialize();
    const config = client.parseConfigFile(realDataset);
    const snapshot = client.updateSnapshot({ openProject: realDataset });
    snapshotHandle = snapshot.snapshot;
    const project = snapshot.projects[0];
    if (!project) {
      throw new Error("real tsgo example did not return a project");
    }

    const primaryFile =
      config.fileNames.find((fileName) => !fileName.endsWith(".d.ts")) ?? config.fileNames[0];
    if (!primaryFile) {
      throw new Error("real tsgo example did not find a source file");
    }

    const sourceFile = client.getSourceFile(snapshot.snapshot, project.id, primaryFile);
    const stringType = client.getStringType(snapshot.snapshot, project.id);

    return {
      fileName: primaryFile,
      projectId: project.id,
      sourceLength: sourceFile?.byteLength ?? 0,
      stringTypeText: client.typeToString(snapshot.snapshot, project.id, stringType.id),
    };
  } finally {
    if (snapshotHandle) {
      client.releaseHandle(snapshotHandle);
    }
    client.close();
  }
}

if (isMain(import.meta.url)) {
  console.log(JSON.stringify(runRealSnapshotExample(), null, 2));
}

File: examples/rust/real_snapshot.rs
mod support;

use corsa_bind_rs::{
    TsgoError,
    api::{ApiClient, ApiMode, UpdateSnapshotParams},
    runtime::block_on,
};
use serde_json::json;

fn main() -> Result<(), corsa_bind_rs::TsgoError> {
    let result = block_on(async {
        let workspace_root = support::workspace_root();
        let dataset = support::real_dataset();
        let dataset_wire = dataset.display().to_string();
        let client = ApiClient::spawn(support::real_api_config(
            "real_snapshot",
            ApiMode::SyncMsgpackStdio,
        )?)
        .await?;
        let init = client.initialize().await?;
        let config = client.parse_config_file(dataset_wire.as_str()).await?;
        let snapshot = client
            .update_snapshot(UpdateSnapshotParams {
                open_project: Some(dataset_wire.clone()),
                file_changes: None,
            })
            .await?;
        let project = snapshot.projects.first().ok_or_else(|| {
            TsgoError::Protocol("real snapshot example did not return a project".into())
        })?;
        let primary_file = config
            .file_names
            .iter()
            .find(|file| !file.ends_with(".d.ts"))
            .cloned()
            .or_else(|| config.file_names.first().cloned())
            .ok_or_else(|| {
                TsgoError::Protocol("real snapshot example did not find a file".into())
            })?;
        let source = client
            .get_source_file(
                snapshot.handle.clone(),
                project.id.clone(),
                primary_file.as_str(),
            )
            .await?
            .ok_or_else(|| {
                TsgoError::Protocol("real snapshot example did not return source".into())
            })?;
        let string_type = client
            .get_string_type(snapshot.handle.clone(), project.id.clone())
            .await?;
        let rendered = client
            .type_to_string(
                snapshot.handle.clone(),
                project.id.clone(),
                string_type.id.clone(),
                None,
                None,
            )
            .await?;

        let result = json!({
            "currentDirectory": support::normalize_path(&workspace_root, &init.current_directory),
            "configFileName": support::normalize_path(&workspace_root, &project.config_file_name),
            "fileCount": config.file_names.len(),
            "primaryFile": support::normalize_path(&workspace_root, primary_file.as_str()),
            "projectId": project.id,
            "sourceLength": source.as_bytes().len(),
            "stringTypeText": rendered,
        });
        snapshot.release().await?;
        client.close().await?;
        Ok::<_, corsa_bind_rs::TsgoError>(result)
    })?;

    support::print_json(result);
    Ok(())
}
